fix: Backstage ESlint rules signifiant speedup

GabDug opened the pull request:
## :wave:

### Context & Problem

While I was trying to improve my ESLint config, I noticed with `TIMING=1` that the slowest rule was taking an order of magnitude more than the other - and that it was a `@backstage/` one!

`TIMING=all yarn backstage-cli repo lint --fix`:
Rule                                                   | Time (ms) | Relative
:------------------------------------------------------|----------:|--------:
@backstage/no-forbidden-package-imports                | 12962.237 |    69.0%
@typescript-eslint/no-redeclare                        |   932.467 |     5.0%
@backstage/no-undeclared-imports                       |   685.183 |     3.6%
@backstage/no-relative-monorepo-imports                |   564.529 |     3.0%
notice/notice                                          |   339.391 |     1.8%
... | ... | ...

### Cause

- `manypkg.getPackagesSync()` is not superfast (scanning files...)
- `clearCache` was called to aggressively in `no-undeclared-import`
- cached only for 5s, while most runs are longer than that

### Solution

- `clearCache` in `no-undeclared-import` only when actions are taken
- Longer caching (5s -> 30s)
  - I believe it's a good balance between general speed and needing to reload ESlint server when working locally 

> In an earlier solution, I replaced `manypkg.getPackagesSync()` with a custom implementation that was faster, but more specialized.
> I believe it will be too much of a maintenance burden, and that performance optimizations be carried upstream (which already supports multiple tools), or by overriding the [scanning tools](https://github.com/Thinkmill/manypkg/tree/main/packages/tools/src).
> The current improvements are already significant, and have less risks.


### Quick benchmark

On the backstage/backstage repository:

**Before:**:
```
Benchmark old: yarn backstage-cli repo lint --fix
  Time (mean ± σ):     32.125 s ±  2.680 s    [User: 120.888 s, System: 38.496 s]
  Range (min … max):   29.555 s … 35.659 s    5 runs
```

Rule                                                   | Time (ms) | Relative
:------------------------------------------------------|----------:|--------:
@backstage/no-forbidden-package-imports                | 12962.237 |    69.0%
@typescript-eslint/no-redeclare                        |   932.467 |     5.0%
@backstage/no-undeclared-imports                       |   685.183 |     3.6%
@backstage/no-relative-monorepo-imports                |   564.529 |     3.0%
notice/notice                                          |   339.391 |     1.8%
... | ... | ...

**After:**
```
Benchmark 1: yarn backstage-cli repo lint --fix
  Time (mean ± σ):     20.434 s ±  1.533 s    [User: 94.206 s, System: 8.612 s]
  Range (min … max):   18.759 s … 22.589 s    5 runs
```

Rule                                                   | Time (ms) | Relative
:------------------------------------------------------|----------:|--------:
@typescript-eslint/no-redeclare                        |  1195.297 |    15.5%
@backstage/no-relative-monorepo-imports                |   612.720 |     8.0%
@backstage/no-undeclared-imports                       |   585.354 |     7.6%



<details><summary>**(Reverted) Custom scanning of workspaces**</summary>
<p>


```
hyperfine --warmup 1  --runs 5 'yarn backstage-cli repo lint --fix'
Benchmark 1: yarn backstage-cli repo lint --fix
  Time (mean ± σ):     17.220 s ±  1.862 s    [User: 86.202 s, System: 7.654 s]
  Range (min … max):   15.660 s … 19.939 s    5 runs
```

Rule                                                   | Time (ms) | Relative
:------------------------------------------------------|----------:|--------:
@typescript-eslint/no-redeclare                        |  1197.588 |    22.7%
... | ... | ...
@backstage/no-forbidden-package-imports                |    30.434 |     0.6%
... | ... | ...
@backstage/no-relative-monorepo-imports                |     3.866 |     0.1%
... | ... | ...
</p>
</details> 

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

drodil commented on packages/eslint-plugin/lib/getPackages.js:
```diff
@@ -45,22 +45,26 @@ module.exports = (function () {
       }
     }
+
+    console.log('dir', dir);
```
Extra logging?

GabDug commented:
> Cool! In general - don't be afraid of using the yaml and globby packages. The robustness is a good tradeoff for the bytes otherwise saved.

@freben To ease maintenance and evolutions, I've reverted the changes with the custom implementation . Even though it could bring more performance,`manypkg` already supports multiple workspace managers with acceptable scanning performance.

Rugvip commented on yarn.lock:
```diff
@@ -4492,7 +4492,7 @@ __metadata:
   dependencies:
     "@backstage/cli": "workspace:^"
-    "@manypkg/get-packages": "npm:^1.1.3"
+    "@manypkg/get-packages": "npm:^2.2.2"
```
Does this fix the original performance issue? I did a quick test locally with `@manypkg/get-packages` and for this repo it's only ~15ms or so to discover all packages.

Rugvip reviewed (commented):
Thank you for looking into this! 👍

GabDug commented on packages/eslint-plugin/lib/getPackages.js:
```diff
@@ -45,21 +45,24 @@ module.exports = (function () {
-      if (Date.now() - lastLoadAt > 5000) {
+      // Cache for 30s
+      // Good balance between too much package scanning and the need to reload the ESLint server in local development
+      if (Date.now() - lastLoadAt > 30000) {
```
Understood!

freben commented:
The test setup has received some fixes lately. A rebase on latest master should help things go green.

drodil commented:
Also a bit related issue: https://github.com/backstage/backstage/issues/30190

freben reviewed (commented):
@GabDug  What is the status of this PR now? Do the two cache clears actually counterintuitively speed things up? Or the bumped packages?

GabDug commented:
@freben I've just rested: with this MR (cache invalidation only when fixes are applied), I go from ~17s to ~9s.

I switched to a more powerful device, so I can't replicate the original benchmarks from the description.

I've just rebased and removed extra changes to keep only the change in cache invalidation.
